### DISCUSSION TITLE
assign vs context (and complex context)

### DISCUSSION THREAD

**Discussion by TimBailey-pnk** [statelyai/xstate#2597] [Q&A]
Hi, I've struggling to understand the `assign` action and why that is prefered to to mutating the context.  In my case I'm using Aurelia to databind the the context and it seems that using assign on deeply nested objects is awkward (I'm missing something for sure)

In the following example I have 2 variants.... in the loaded state I have some entry actions.... if I use assign then the updated context is available in the guards, but if I just use a function then it is not....
```js
const listMachine = createMachine({
  id: 'transactionlist',
  initial: 'uninitialized',
  context,
  states: {
    uninitialized: {
      on: {
        LISTOPTIONSCHANGED: {
          target: 'loaded'
        }
      }
    },
    loaded: {
      entry: [
        log('running entry'),
        assign({
          a: 1
        }),
        // (context, event) => { <---- swap this action with the above action
        //   context.a = 2;
        // }
      ],
      always: [
        { cond: 'hasA', target: 'a' },
        { cond: 'hasB', target: 'b' }
      ]
    },
    a: {},
    b: {}
  }
}, 
{
  guards: {
    hasA: (context, event) => {
      console.log('context.a', context.a)
      return context.a
    },
    hasB: (context, event) => {
      console.log('context.b', context.b)
      return context.b
    },
  }
})
```
for assign I get this output
> guard for context.a 1
> running entry

for the mutate I get this output
> guard for context.a undefined
> guard for context.b undefined
> running entry

So its running the assign for the entry before the guards, should the guards be evaluated after the entry?
And I see that using assign is the recommended approach....

Any suggestions for using assign to update a child object in a complex context, ie somthing like
```
{
   currentItem: {
     a: 1,
     b: 2
  }
}
```
I want to assign a=3 and retain the other properties and the reference to currentItem?

Thanks

**Comment by davidkpiano** [comment]
Non-built-in actions are executed _after_ the state is resolved, which is why the guards don't see the updated context.

To prevent mutation, it's always recommended to update `context` through an `assign(...)` action. Otherwise, safety is lost if `context` is allowed to be mutated anywhere at any time.

**Reply by TimBailey-pnk** [reply]
Thanks, am I modelling incorrectly, should I be aiming for simple (ie not nested) context?

**Reply by davidkpiano** [reply]
Your context can be as nested as you want, and you can always assign to it by using a function:

```js
actions: assign((context) => {
  // immutably return the next context
  // ...
});
```

The shorthands are for shallow updates to context (`assign({ one: 1, two: (ctx) => ctx.one + 1 })`).

**Reply by TimBailey-pnk** [reply]
ok, i think I've finally got whats messing me up, I'm finding the "rollup" of assigns confusing, not in the actions array for a single event descriptor, but when I have a entry, exit and other actions, then the assigns are out of order for even entry etc.

Checking the notes you mention this behaviour will change in V5 of xstate?  And then will run in order?  Any idea of date for that even a beta version?

**Reply by ChrisShank** [reply]
There is backwards compatibility for that feature already released in v4! https://github.com/statelyai/xstate/blob/main/packages/core/CHANGELOG.md#minor-changes

**Reply by TimBailey-pnk** [reply]
Thanks!!!  Checking that now
